src/lashup_kv.erl

```diff
@@ -229,72 +233,114 @@ init_db(Nodes) ->
+        ok; %% already upgraded
+    {false, false} ->
+        ok; %% nothing to upgrade
+    {false, true} ->
+        upgrade_table(kv)
+  end.
+
+-spec(upgrade_table(kv) -> ok|aborted).
+upgrade_table(kv) ->
+  F = fun() ->
+        Records = mnesia:foldl(fun convert_and_write_record/2, [], ?OLD_KV_TABLE),
+        NClock = #nclock{key = node(), lclock = length(Records)},
+        mnesia:write(NClock)
+      end,
+  case mnesia:sync_transaction(F) of
+    {atomic, _} ->
+      ok = mnesia:sync_log();
+    {aborted, Reason} ->
+      lager:error("Failed to upgrade kv table because ~p", [Reason]),
+      aborted
+  end.
+
+convert_and_write_record(Record0, Acc) ->
+    Record1 = convert_record(Record0, length(Acc)),
+    mnesia:write(Record1),
+    [Record1 | Acc].
+
+convert_record(_R = #kv{key = Key, vclock = VClock, map = Map}, Counter) ->
+  #kv2{key = Key, vclock = VClock, map = Map, lclock = Counter}.
+
+-spec(mk_write_fun(Key :: key(), OldVClock :: riak_dt_vclock:vclock() | undefined,
+      Op :: riak_dt_map:map_op()) -> (fun())).
 mk_write_fun(Key, OldVClock, Op) ->
-  Node = node(),
   fun() ->
-    NewKV =
-      case mnesia:read(kv, Key, write) of
+    {NewKV, NClock} =
+      case mnesia:read(?KV_TABLE, Key, write) of
         [] ->
-          VClock = riak_dt_vclock:increment(Node, riak_dt_vclock:fresh()),
-          Counter = riak_dt_vclock:get_counter(Node, VClock),
-          Dot = {Node, Counter},
-          {ok, Map} = riak_dt_map:update(Op, Dot, riak_dt_map:new()),
-          #kv{key = Key, vclock = VClock, map = Map};
-        [_ExistingKV = #kv{vclock = VClock}] when OldVClock =/= undefined andalso VClock =/= OldVClock ->
+          prepare_kv(Key, riak_dt_map:new(), riak_dt_vclock:fresh(), Op);
+        [#kv2{vclock = VClock}] when OldVClock =/= undefined andalso VClock =/= OldVClock ->
           mnesia:abort(concurrency_violation);
-        [ExistingKV = #kv{vclock = VClock, map = Map}] ->
-          VClock1 = riak_dt_vclock:increment(Node, VClock),
-          Counter = riak_dt_vclock:get_counter(Node, VClock1),
-          Dot = {Node, Counter},
-          {ok, Map1} = riak_dt_map:update(Op, Dot, Map),
-          ExistingKV#kv{vclock = VClock1, map = Map1}
+        [#kv2{vclock = VClock, map = Map}] ->
+          prepare_kv(Key, Map, VClock, Op)
       end,
     case check_map(NewKV) of
       {error, Error} ->
         mnesia:abort(Error);
       ok ->
-        mnesia:write(NewKV)
+        mnesia:write(NewKV),
+        mnesia:write(NClock)
     end,
     NewKV
   end.
 
+-spec(prepare_kv(Key :: key(), Map0 :: riak_dt_map:dt_map(), VClock0 :: riak_dt_vclock:vclock() | undefined,
+      Op :: riak_dt_map:map_op()) -> {kv(), nclock()}).
+prepare_kv(Key, Map0, VClock0, Op) ->
+  Node = node(),
+  VClock1 = riak_dt_vclock:increment(Node, VClock0),
+  Counter = riak_dt_vclock:get_counter(Node, VClock1),
+  Dot = {Node, Counter},
+  {ok, Map1} = riak_dt_map:update(Op, Dot, Map0),
+  LClock0 = get_clock(Node),
```

Comment:
One final thing on the get_clock function, and increment, maybe get_lclock, and increment_lclock -- just something to think about?